kinto_fxa/authentication.py

```diff
@@ -85,29 +94,39 @@ def _verify_token(self, token, request):
         # `request.bound_data` is an attribute provided by Kinto to store
         # some data that is shared among sub-requests (e.g. default bucket
         # or batch requests)
-        key = 'fxa_verified_token'
-        if key in request.bound_data:
-            return request.bound_data[key]
-
-        # Use PyFxa defaults if not specified
-        server_url = fxa_conf(request, 'oauth_uri')
-        scope = aslist(fxa_conf(request, 'required_scope'))
-        auth_cache = self._get_cache(request)
-        auth_client = OAuthClient(server_url=server_url, cache=auth_cache)
-        try:
-            profile = auth_client.verify_token(token=token, scope=scope)
-            user_id = profile['user']
-        except fxa_errors.OutOfProtocolError as e:
-            logger.exception("Protocol error")
-            raise httpexceptions.HTTPServiceUnavailable()
-        except (fxa_errors.InProtocolError, fxa_errors.TrustError) as e:
-            logger.debug("Invalid FxA token: %s" % e)
+        if KEY not in request.bound_data:
+            # Use PyFxa defaults if not specified
+            server_url = fxa_conf(request, 'oauth_uri')
+            auth_cache = self._get_cache(request)
+            auth_client = OAuthClient(server_url=server_url, cache=auth_cache)
+
             user_id = None
+            client_name = None
+
+            for scope, client in request.registry._fxa_oauth_scope_routing.items():
+                try:
+                    profile = auth_client.verify_token(token=token, scope=aslist(scope))
+                    user_id = profile['user']
+                    scopes = profile['scope']
+                    client_name = client
+
+                    # Make sure the bearer token scopes don't match multiple configs.
+                    if len([x for x in request.registry._fxa_oauth_scope_routing.keys()
+                            if x and set(x.split()) & set(scopes)]) > 1:
```

Comment:
It's not clear to me whether this is doing what you intend.  What if the scope routing here was something like:

```
{
  "profile https://identity.mozilla.com/app/notes": "notes",
  "profile https://identity.mozilla.com/app/lockbox": "lockbox",
}
```

Then it seems to me that a token with "profile https://identity.mozilla.com/app/lockbox" would also match the set-intersection check above for "profile https://identity.mozilla.com/app/notes", because they both contain the "profile" scope.  Is that the intended behaviour?

Or, if the keys in `_fxa_oauth_scope_routing` are only ever intended to be a single scope like "https://identity.mozilla.com/app/notes" rather than a list of scopes, maybe we should explicitly error out rather than doing `set(x.split())` and so-on here.

On a side note: there's a lot of string splitting and parsing being done here on each request, I wonder if it would make sense to parse the scopes into a `set` or `frozenset` during config loading to make the checking here quicker.

Reply:
Yes I need to add a few tests for that.